150 create grpc overview feature page (#170)
aklivity/zilla-docs: src/concepts/kafka-proxies/grpc-proxy.md

---
description: The Zilla gRPC Kafka Proxy lets you implement gRPC service definitions from protobuf files to consume and produce messages from Kafka topics.
prev: false
next: /tutorials/grpc/grpc-intro.md
---

# gRPC Kafka Proxy

> The Zilla gRPC Kafka Proxy lets you implement gRPC service definitions from protobuf files to consume and produce messages from Kafka topics.

Reviewer: ```suggestion
The Zilla gRPC Kafka Proxy lets you implement gRPC service definitions from protobuf files to produce and consume messages via Kafka topics.
```
seems more logical to produce before consume, especially for request-response, and via is neutral to both directions, whereas to and from would be direction-specific

> A service method's request and response messages are routed on and off of Kafka. Zilla can be the server delivering messages to a Kafka topic or fanout messages from a topic to running gRPC services. Additionally, Zilla can sit on the critical path between a gRPC client and server. They can communicate as if they are talking to each other, with Zilla proxying the messages through Kafka.

Reviewer: ```suggestion
Zilla can be the gRPC server, routing a service method's request and response messages to and from Kafka topics, or Zilla can fanout messages from a Kafka topic to multiple gRPC clients. Additionally, Zilla can sit on the critical path between a gRPC client and server. They can communicate as if they are talking directly to each other, while Zilla actually proxies the messages through Kafka.
```

## Correlated Request-Response

> Zilla manages the synchronous request and response messages of a gRPC service on the event stream. Each message is correlated to each other with a `zilla:correlation-id` header, providing an identifier for both Zilla and Kafka workflows to act on. Correlated messages can be on the same or different Kafka topics.

Reviewer: The concept of event stream in Kafka is typically used for facts (events) that have already happened, so this is better aligned terminology for fanout, whereas request-response is more like command-reply pattern than events, since commands are a request to do work and can be rejected.

> Each message is correlated to each other with a ...

Request and responses messages are correlated by a ...

> Correlated messages can be on the same or different Kafka topics.

Correlated messages can really only be on different topics.
The only exception to this is doing some kinds of echo test where the request is treated immediately as the response, and is therefore implicitly correlated because it is the same message with the same correlation header value.
In practice this is not useful, so we should only emphasize the need for 2 topics here, one for requests and one for responses.

## RPC Service Definitions

> Zilla supports all four [gRPC service method definitions](https://grpc.io/docs/what-is-grpc/core-concepts/#service-definition). The request messages are routed to a Kafka topic. The return message(s) can be delivered to the same or a different topic. Zilla can also handle the stream upgrade when a client sends a single request, but the service expects a stream.

Reviewer: > The return message(s) can be delivered to the same or a different topic.

The return message(s) are delivered to a different topic.

> - **Simple/Unary RPC** - A single message is sent and will wait for the correlated response message and return it back to the caller.

Reviewer: This is describing what happens at gRPC side, but I think we need to cover what is happening in Kafka too.

We send a Kafka message representing the gRPC request message, then send a second null Kafka message to represent the end of the request. We follow the same approach for the response too.

  ```protobuf:no-line-numbers
  rpc SayHello(HelloRequest) returns (HelloResponse);
  ```

> - **Server-side streaming RPC** - A single message is sent with a returned stream back to the caller. The correlated messages produced on the reply-to topic will be sent for the client to read until there are no more messages, and the stream will close.

Reviewer: Suggest we mention Kafka representation of end-of-stream (null message) here as well.

  ```protobuf:no-line-numbers
  rpc LotsOfReplies(HelloRequest) returns (stream HelloResponse);
  ```

> - **Client-side streaming RPC** - The client sends a stream, producing all the messages on a topic. When the client finishes writing to the stream, it will wait for the correlated response message and return it back to the caller.

Reviewer: Suggest we mention Kafka representation of end-of-stream (null message) here as well.

  ```protobuf:no-line-numbers
  rpc LotsOfGreetings(stream HelloRequest) returns (HelloResponse);
  ```

> - **Bidirectional streaming RPC** - Both the client and server use a read-write stream to produce and consume correlated messages.

Reviewer: Suggest we mention Kafka representation of end-of-stream (null message) here as well.

  ```protobuf:no-line-numbers
  rpc BidiHello(stream HelloRequest) returns (stream HelloResponse);
  ```

## gRPC Metadata

> Custom metadata fields can be used for request routing and idempotency. Metadata is preserved through Kafka for both the request and response messages. Zilla can augment the metadata it sends based on the configured route the request matches.

Reviewer: > Metadata is preserved through Kafka for both the request and response messages.

We verified this is not the case, agree?

Author: agreed, I moved the sentence to a different branch so we can merge when the feature is released.

## RPC Service Definitions

Zilla supports all four [gRPC service method definitions](https://grpc.io/docs/what-is-grpc/core-concepts/#service-definition). The request and return message(s) are managed through two different Kafka topics, respectively. These topics are defined through [dynamic method routing](../../concepts/config-intro.md#routes).

Zilla can also handle the stream upgrade when a client sends a single request, but the service expects a stream. Zilla does this by treating all gRPC request and response messages as a stream of messages on Kafka topics with at least one data message and a null end-of-stream message representing the end of the request or response streams.

### Simple/Unary RPC

```protobuf:no-line-numbers
rpc SayHello(HelloRequest) returns (HelloResponse);
```

A single message is sent and will wait for the correlated response message and return it back to the caller. The request and response topics both have one message with the method payloads and one end-of-stream message.

### Server-side streaming RPC

```protobuf:no-line-numbers
rpc LotsOfReplies(HelloRequest) returns (stream HelloResponse);
```

> A single message is sent with a returned stream back to the caller. The correlated messages produced on the reply-to topic will be sent for the client to read until there are no more messages, and the stream will close. The request topic has one message with the method payloads and one end-of-stream message. The response topic has one or many messages with the method payloads and one end-of-stream message.

Reviewer: ```suggestion
A single message is sent with a returned stream back to the caller. The correlated messages produced on the reply-to topic will be sent for the client to read until there are no more messages, and the stream will close. The request topic has one message with the method payload and one end-of-stream message. The response topic has one or many messages with the method payloads and one end-of-stream message.
```

### Client-side streaming RPC

```protobuf:no-line-numbers
rpc LotsOfGreetings(stream HelloRequest) returns (HelloResponse);
```

> The client sends a stream, producing all the messages on a topic. When the client finishes writing to the stream, it will wait for the correlated response message and return it back to the caller. The request topic has one or many messages with the method payloads and one end-of-stream message. The response topic has one message with the method payloads and one end-of-stream message.

Reviewer: It's possible that the correlated response message can be sent before the request stream has completed.

### Bidirectional streaming RPC

```protobuf:no-line-numbers
rpc BidiHello(stream HelloRequest) returns (stream HelloResponse);
```

Both the client and server use a read-write stream to produce and consume correlated messages. The request and response topics have one or many messages with the method payloads and one end-of-stream message.

## Correlated Request-Response

Zilla manages the synchronous request and response messages of a gRPC service. Request and response messages are correlated by a `zilla:correlation-id` header, providing an identifier for both Zilla and Kafka workflows to act on.

## gRPC Metadata

Method routes can use [custom metadata fields](../../reference/config/bindings/binding-grpc-kafka.md#when-metadata) for request routing and idempotency. Zilla can also augment the metadata it sends based on the configured route the request matches.

## Reliable Delivery

> Zilla sends an event ID with every message serialized as an unknown field in the payload. Any message can be identified without field collision, and the client doesn't need to acknowledge the message receipt explicitly. A client consuming a stream of messages can remember the event ID. If the event the stream gets interrupted. The client reconnects with a `last-event-id` header to recover without message loss or needing to start over from the beginning.

Reviewer: I think this last section feels like it is part of the other sections, like you can still do all those things above but also add reliable delivery.

However, this reliably delivery is a oneway broadcast only, so I think we need to make that context more clear before describing the reliability aspects, just to avoid any potential confusion.
